src/averageAges.js

```diff
@@ -1,61 +1,86 @@
+    const onlyMothersList = [];
+
+    people.forEach(person => {
+      if (typeof person.mother === 'string') {
+        onlyMothersList.push(person.mother);
+      }
+    });
+
+    const uniqueMothersList = [...new Set(onlyMothersList)];
+    const mothers = uniqueMothersList.map(motherName => {
+      return people.find(person => person.name === motherName);
+    });
+
+    return mothers.filter(mother => typeof mother !== 'undefined');
+  },
+
+  getSumAgeDiffWithMothers(sons, mothers) {
+    return sons.reduce((result, son) => result + mothers[son.name], 0);
+  },
+};
+
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = methods.getFilteredPeopleBySex(people, 'm');
+  const menByCentury = century >= 1
+    ? methods.getFilteredMenByCentury(men, century)
+    : men;
+
+  return methods.getAverage(menByCentury);
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const onlyMothersList = methods.getOnlyMothersList(people);
+  const allWomenList = methods.getFilteredPeopleBySex(people, 'f');
+
+  return withChildren
+    ? methods.getAverage(onlyMothersList)
+    : methods.getAverage(allWomenList);
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const mothers = people.reduce((result, person) => {
+    const mother = typeof person.mother === 'string'
+    && people.find(woman => woman.name === person.mother);
+
+    mother && (result[person.name] = person.born - mother.born);
+
+    return result;
+  }, {});
+  const children
+  = people.filter(person => mothers.hasOwnProperty(person.name));
+  const sons = methods.getFilteredPeopleBySex(children, 'm');
+  const totalAgeDiff = onlyWithSon
+    ? methods.getSumAgeDiffWithMothers(sons, mothers)
+    : methods.getSumAgeDiffWithMothers(children, mothers);
+  const count = onlyWithSon
+    ? sons.length
+    : children.length;
+
+  return totalAgeDiff / count;
```

Comment:
Hard to read and understand. Simplify logic, use less repeated code, combine conditions inside filter. 3 simple steps:
- filter all mothers in 1 `filter` call;
- calculate total age diff using 1 `reduce` call (for each woman find her child just inside reduce);
- return average;